// apollo/examples/counter/worker/index.js
var Service = require('./service');

exports.children = {
    'service': Service
};

exports.events = {
    'start': 'initialize'
  , 'count': 'count'
};

exports.receive = {
    'start': 'start'
};

exports.initialize = function (ev, done) {
  this.service = this.ctx.startChild('service');
  this.maxCount = 51;
  this.progressListener = null;
  this.interval = null;
  done();
};

exports.start = function (msg, done) {
  var self = this;

  if (!this.progressListener) {
    console.log('starting worker!');
    this.progressListener = msg.sender;
    this.interval = setInterval(function () {
      self.emit('count');
    }, 1000);
  }

  done();
};

exports.count = function () {
  var self = this
    , cs = this.service
    , pl = this.progressListener
    , max = this.maxCount;

  cs.tell('increment')
    .set('value', 1);

  cs.tell('increment')
    .set('value', 1);

  cs.tell('increment')
    .set('value', 1);

  cs.ask('status')
    .res(function (count) {
      var progress = Math.floor(100 * (count/ max));
      if (progress >= 100) clearInterval(self.interval);
      pl.tell('progress')
        .set('value', progress);
    });
};
